File: DSA/6. Graph/16_Prim_Algo.cpp
#include <bits/stdc++.h>
using namespace std;

int spanningTree(int v, vector<vector<int>> adj[])
{
    int mincost = 0;
    vector<int> costs(v, INT_MAX);

    costs[0] = 0;
    vector<bool> visited(v, false);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    pq.push({0, 0});

    while (!pq.empty())
    {
        pair<int, int> p = pq.top();
        int currnode = p.second;
        int currcost = p.first;
        pq.pop();

        if (visited[currnode])
            continue;

        mincost += currcost;
        visited[currnode] = true;
        costs[currnode] = currcost;

        for (int i = 0; i < adj[currnode].size(); i++)
        {
            int nextnode = adj[currnode][i][0];
            int nextcost = adj[currnode][i][1];
            if (visited[nextnode])
                continue;

            pq.push({nextcost, nextnode});
        }
    }
    return mincost;
}

File: DSA/6. Graph/16_Prim_Algo.java
/*
    Problem Link: https://www.geeksforgeeks.org/prims-minimum-spanning-tree-mst-greedy-algo-5/
    
    Prim's Algorithm - Minimum Spanning Tree
*/

import java.util.*;

class Prim_Algo {
    public static int spanningTree(int v, List<int[]>[] adj) {
        int mincost = 0;
        int[] costs = new int[v];
        Arrays.fill(costs, Integer.MAX_VALUE);

        costs[0] = 0;
        boolean[] visited = new boolean[v];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        pq.offer(new int[]{0, 0});

        while (!pq.isEmpty()) {
            int[] p = pq.poll();
            int currnode = p[1];
            int currcost = p[0];

            if (visited[currnode]) {
                continue;
            }

            mincost += currcost;
            visited[currnode] = true;
            costs[currnode] = currcost;

            for (int i = 0; i < adj[currnode].size(); i++) {
                int nextnode = adj[currnode].get(i)[0];
                int nextcost = adj[currnode].get(i)[1];
                if (visited[nextnode]) {
                    continue;
                }

                pq.offer(new int[]{nextcost, nextnode});
            }
        }
        return mincost;
    }
}
